src/node.cc

```diff
@@ -1299,86 +1299,187 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
-      // NOTE: For backwards compatibility with public API we return Undefined()
-      // if the top level call threw.
-      return callback_scope.in_makecallback() ?
-          ret : Undefined(env->isolate());
-    }
+  if (try_catch_ != nullptr && try_catch_->real_try_catch.HasCaught())
+    failed_ = true;
 
-    if (asyncContext.async_id != 0) {
-      AsyncWrap::EmitAfter(env, asyncContext.async_id);
-    }
+  delete try_catch_;
+  try_catch_ = nullptr;
+
+  if (pushed_ids_)
+    env_->async_hooks()->pop_ids(async_context_.async_id);
+
+  if (failed_) return;
+
+  if (async_context_.async_id != 0) {
+    AsyncWrap::EmitAfter(env_, async_context_.async_id);
   }
 
-  if (env->using_domains()) {
-    disposed_domain = DomainExit(env, object);
-    if (disposed_domain) return Undefined(env->isolate());
+  if (env_->using_domains()) {
+    failed_ = DomainExit(env_, object_);
+    if (failed_) {
+      failure_return_value_ = Undefined(env_->isolate());
+      return;
+    }
   }
 
-  if (callback_scope.in_makecallback()) {
-    return ret;
+  if (IsInnerMakeCallback()) {
+    return;
   }
 
-  Environment::TickInfo* tick_info = env->tick_info();
+  Environment::TickInfo* tick_info = env_->tick_info();
 
   if (tick_info->length() == 0) {
-    env->isolate()->RunMicrotasks();
+    env_->isolate()->RunMicrotasks();
   }
 
   // Make sure the stack unwound properly. If there are nested MakeCallback's
   // then it should return early and not reach this code.
-  CHECK_EQ(env->current_async_id(), asyncContext.async_id);
-  CHECK_EQ(env->trigger_id(), asyncContext.trigger_async_id);
+  CHECK_EQ(env_->current_async_id(), 0);
+  CHECK_EQ(env_->trigger_id(), 0);
 
-  Local<Object> process = env->process_object();
+  Local<Object> process = env_->process_object();
 
   if (tick_info->length() == 0) {
     tick_info->set_index(0);
-    return ret;
+    return;
+  }
+
+  CHECK_EQ(env_->current_async_id(), 0);
+  CHECK_EQ(env_->trigger_id(), 0);
+
+  if (env_->tick_callback_function()->Call(process, 0, nullptr).IsEmpty()) {
+    failed_ = true;
+    failure_return_value_ = Undefined(env_->isolate());
+  }
+}
+
+MaybeLocal<Value> InternalMakeCallback(Environment* env,
+                                       Local<Value> recv,
+                                       const Local<Function> callback,
+                                       int argc,
+                                       Local<Value> argv[],
+                                       async_context asyncContext) {
+  InternalCallbackScope scope(env, recv, asyncContext);
+  if (scope.Failed()) {
+    return scope.GetFailureReturnValue();
+  }
+
+  MaybeLocal<Value> ret;
+
+  {
+    ret = callback->Call(env->context(), recv, argc, argv);
+
+    if (ret.IsEmpty()) {
+      // NOTE: For backwards compatibility with public API we return Undefined()
+      // if the top level call threw.
+      scope.MarkAsFailed();
+      return scope.IsInnerMakeCallback() ? ret : Undefined(env->isolate());
```

Comment:
> The only time `Undefined()` is returned is in the case of `InternalMakeCallback()` being wrapped in a `TryCatch`, and is not rethrown.

Why is that?

> It may be considered a semver-major, but looks like always returning `ret` would simplify code elsewhere.

Yes, `MakeCallback` should always return an empty MaybeLocal in case of a thrown exception, but yes, that would be up to a semver-major follow-up PR.